src/aggregator/client/writer.go

```diff
@@ -89,70 +99,82 @@ func newInstanceWriter(instance placement.Instance, opts Options) instanceWriter
 		queue:             newInstanceQueue(instance, queueOpts),
 		encodersByShard:   make(map[uint32]*lockedEncoder),
 	}
+
+	w.isFlushActive.Store(false)
 	w.newLockedEncoderFn = newLockedEncoder
 	return w
 }
 
+// Write takes the payload and writes it to the per-shard encoder.
 func (w *writer) Write(shard uint32, payload payloadUnion) error {
-	w.RLock()
-	if w.closed {
-		w.RUnlock()
-		return errInstanceWriterClosed
+	w.encoderByShardMtx.RLock()
+	if w.closed.Load() {
+		w.encoderByShardMtx.RUnlock()
+		return ErrInstanceWriterClosed
 	}
 	encoder, exists := w.encodersByShard[shard]
 	if exists {
 		err := w.encodeWithLock(encoder, payload)
-		w.RUnlock()
+		w.encoderByShardMtx.RUnlock()
 		return err
 	}
-	w.RUnlock()
+	w.encoderByShardMtx.RUnlock()
 
-	w.Lock()
-	if w.closed {
-		w.Unlock()
-		return errInstanceWriterClosed
+	w.encoderByShardMtx.Lock()
+	if w.closed.Load() {
+		w.encoderByShardMtx.Unlock()
+		return ErrInstanceWriterClosed
 	}
 	encoder, exists = w.encodersByShard[shard]
 	if exists {
 		err := w.encodeWithLock(encoder, payload)
-		w.Unlock()
+		w.encoderByShardMtx.Unlock()
 		return err
 	}
 	encoder = w.newLockedEncoderFn(w.encoderOpts)
 	w.encodersByShard[shard] = encoder
 	err := w.encodeWithLock(encoder, payload)
-	w.Unlock()
+	w.encoderByShardMtx.Unlock()
 
 	return err
 }
 
+// Flush loops through all encoders in encodersByShard, grabs the encoded
+// payload and writes it to the queue. From there the queue is drained
+// towards the destination via the transport in a blocking manner
 func (w *writer) Flush() error {
-	w.RLock()
-	if w.closed {
-		w.RUnlock()
-		return errInstanceWriterClosed
+	if w.closed.Load() {
+		return ErrInstanceWriterClosed
 	}
-	err := w.flushWithLock()
-	w.RUnlock()
 
-	if err != nil {
+	if !w.isFlushActive.CAS(false, true) {
+		// Flush is already active, bail
+		w.metrics.skippedFlushes.Inc(1)
+		return ErrFlushInProgress
+	}
+
+	defer w.isFlushActive.Store(false)
+
+	if err := w.flush(); err != nil {
 		w.metrics.flushErrors.Inc(1)
 		return err
 	}
+
 	return nil
 }
 
 func (w *writer) Close() error {
-	w.Lock()
-	defer w.Unlock()
-
-	if w.closed {
-		return errInstanceWriterClosed
+	if !w.closed.CAS(false, true) {
+		return ErrInstanceWriterClosed
 	}
-	w.closed = true
-	if err := w.flushWithLock(); err != nil {
-		w.metrics.flushErrors.Inc(1)
+
+	if w.isFlushActive.CAS(false, true) {
```

Comment:
You may have to busy-loop until retrying until the current flush is complete (instead of not doing a flush at all if one is in progress) to catch data added between
``
w.encoderByShardMtx.RUnlock()
...
w.queue.Flush()
```
However I'm not sure how much of that matters if the writer is closing because of a placement change - maybe it's fine to lose a bit of data.

Reply:
True. The only reasons I can think of for Close()ing a writer are placement changes and connection errors. In both cases it seems like we would not need the remaining data. Is that a fair assessment ?

Reply:
Placement changes could happen because of graceful cluster expansion (not just node dying), we don't want to lose data in those cases.